📝 EID-1943 Sign eIDAS metadata with correct algorithms

idrop opened the pull request:
https://govukverify.atlassian.net/browse/EID-1943

The [eIDAS Cryptographic Requirement v.1.2 Final.pdf](https://github.com/alphagov/verify-metadata-controller/files/4409341/eIDAS.Cryptographic.Requirement.v.1.2.Final.pdf) indicates that metadata should be signed with RSASSA-PSS or ECDSA.

CloudHSM keys are [now created with EC secp384r1 algorithm](https://github.com/alphagov/verify-metadata-controller/pull/79), so metadata signed with these keys will use ECDSA.

Existing RSA keys in CloudHSM used to sign metadata will use RSASSA-PSS.

Metadata using ECDSA for signing has been created in sanbox, see  `gds sandbox k -n sandbox-metadata-controller get secret sandbox-metadata-a2 -o yaml | yq -r '.data["metadata.xml"]' | base64 -D`

💥Also, the VMC is now a multi-module gradle project, and has tests! 🙀. This was driven by the need to use `CloudHSMWrapper` from cloudhsmtool in the mdgen module.

💥Plus, certificates created in the CertificateController will have a subject identifier and an authority identifier - see `CertificateRequest` kubeyaml for a new root CA with `CN=Verify Test Root CA A2` that will be built with a subject identifier.

💥Use amazoncorretto as base docker image, and remove gradle build artifacts from final image

idrop committed:
Use gradle on project

Allows us to share CloudHSMWrapper between cloudhsmtool and mdgen.
Allows better project level dependency management
Easier to work on the project in an IDE like IntelliJ
Add unit tests!
Dockerfile build of java CLIs is consolidated
Dockerfile build on amazoncorretto

idrop committed:
Sign metadata with correct algo

Sign metadata based on public key held in HSM
If public key is EC then sign with ecdsa-sha384 algo
If public key is RSSA then sign with sha384-rsa-MGF1 algo

idrop committed:
Create certificates with a subject identifier

Also reference the authority identifier if this certificate is issued by an authority

idrop committed:
When signing metadata with RSA keys, use sha384-rsa-MGF1 algo

JohnWattsGDS commented on gradlew.bat:
```diff
@@ -0,0 +1,103 @@
+@rem
```
Strictly speaking we could delete this file.  It occurs to me that I think I never have in similar situations, so definitely not a deal breaker.

JohnWattsGDS commented on mdgen/src/main/java/uk/gov/ida/mdgen/MetadataGenerator.java:
```diff
@@ -184,28 +183,10 @@ private void setupKeyInfoGeneratorFactory() {
-        LOG.info("Cavium keystore aliases: \\n" + stringBuilder.toString());
-
-        PrivateKey key = (PrivateKey) cloudHsmStore.getKey(label, null);
+        PrivateKey key = cloudHSMWrapper.getPrivateKey(label);
```
🎉

JohnWattsGDS reviewed (dismissed):
👍

CharlesIC commented on mdgen/src/test/java/uk/gov/ida/mdgen/MdgenTestUtils.java:
```diff
@@ -0,0 +1,61 @@
+package uk.gov.ida.mdgen;
```
I wonder if the classpath here should be `uk.gov.ida.mdgen.test`

CharlesIC reviewed (commented):
We should remove `gradlew`, `gradlew.bat` and `setting.gradle` from `mdgen` as it's no longer a root project.

Plus a few questions below

idrop committed:
Remove root project gradle config from modules

idrop commented on chart/templates/cert-request-test-a2.yaml:
```diff
@@ -0,0 +1,20 @@
+  namespace: {{ .Release.Namespace }}
+spec:
+  countryCode: GB
+  commonName: Verify Test Root CA A2
```
It's a distinct new root CA certificate for testing. 
The current roots CA for testing and production on the verify cluster are:
`verify-root-ca-production`
`verify-root-ca-test`
The names above correspond to old *RSA* keys in the HSM which are out of spec.
If you delete the secret derived from this `CertificateRequest` it will use the existing RSA key in the HSM.
To create an EC key we need to create a new `name` in the CertificateRequest.

I created an A1 CertificateRequest previously which created an HSM EC key with label `verify-root-ca-test-a1`, but the certificate did not have a subject identifier. I created A2 to generate a new clean cert with a subject identifier. I'm going to delete this file however, as if we delete the `verify-root-ca-test-a1` secret, the certificate should be recreated with a subject identifier.

idrop replied:
Discussed, with Karol - I'm going to delete this CertificateRequest and re-use CertificateRequest verify-root-ca-test-a1

CharlesIC replied:
Okay thanks for the explanation Phil 👍

idrop committed:
Remove root CA a2, we can regenerate a cert for a1

We will delete the secret `verify-root-ca-test-a1`, then restart the vmc pod.
This will recreate the a1 certificate with a subject identifier and better common name, inside a new verify-root-ca-test-a1 secret.

idrop commented on Dockerfile:
```diff
@@ -23,40 +23,40 @@ RUN yum install -y wget tar gzip openssl \
+
+RUN mv /buildjava/cloudhsmtool/build/install/cloudhsmtool /cloudhsmtool
+RUN mv /buildjava/mdgen/build/install/mdgen /mdgen
+RUN rm -r /buildjava
```
I'm going to keep it as is in that case.

idrop committed:
run travis build from root directory
